dlt/extract/extract.py

```diff
@@ -367,14 +377,44 @@ def extract(
         load_id = self.extract_storage.create_load_package(source.discover_schema())
         with Container().injectable_context(
             SourceSchemaInjectableContext(source.schema)
-        ), Container().injectable_context(SourceInjectableContext(source)):
+        ), Container().injectable_context(
+            SourceInjectableContext(source)
+        ), Container().injectable_context(
+            LoadPackageStateInjectableContext(
+                load_id=load_id, storage=self.extract_storage.new_packages
+            )
+        ) as load_package:
             # inject the config section with the current source name
             with inject_section(
                 ConfigSectionContext(
                     sections=(known_sections.SOURCES, source.section, source.name),
                     source_state_key=source.name,
                 )
             ):
+                if self.refresh is not None:
+                    _resources_to_drop = (
+                        list(source.resources.extracted) if self.refresh != "drop_dataset" else []
+                    )
+                    _state, _ = pipeline_state(Container())
+                    new_schema, new_state, drop_info = drop_resources(
+                        source.schema,
+                        _state,
+                        resources=_resources_to_drop,
+                        drop_all=self.refresh == "drop_dataset",
+                        state_only=self.refresh == "drop_data",
+                        state_paths="*" if self.refresh == "drop_dataset" else [],
+                    )
+                    _state.update(new_state)
+                    if drop_info["tables"]:
+                        drop_tables = [
```

Comment:
we need to drop tables from a full table chain. look what happens in load. also we have a marker (seen-data) that tells which tables were really created.

same thing for truncated tables. we should truncate table chains.

maybe when we have table chains we do not really need to modify schema in load.py

Reply:
@rudolfix  I'm not 100% what you mean. (aside from checking seen-data) the table chain is already generated in drop far as I can tell, the `dropped_tables` list is the result of:

```python
def group_tables_by_resource(
    tables: TSchemaTables, pattern: Optional[REPattern] = None
) -> Dict[str, List[TTableSchema]]:
    """Create a dict of resources and their associated tables and descendant tables
    If `pattern` is supplied, the result is filtered to only resource names matching the pattern.
    """
    result: Dict[str, List[TTableSchema]] = {}
    for table in tables.values():
        resource = table.get("resource")
        if resource and (pattern is None or pattern.match(resource)):
            resource_tables = result.setdefault(resource, [])
            resource_tables.extend(get_child_tables(tables, table["name"]))
    return result

...
data_tables = {t["name"]: t for t in schema.data_tables()}  # could probably include seen_data_only here?
resource_tables = group_tables_by_resource(data_tables, pattern=resource_pattern)
```

Reply:
@steinitzu I overlooked this. we are interested in root tables that seen data so
```
def _write_empty_files(
        self, source: DltSource, extractors: Dict[TDataItemFormat, Extractor]
    ) -> None:
        schema = source.schema
        json_extractor = extractors["object"]
        resources_with_items = set().union(*[e.resources_with_items for e in extractors.values()])
        # find REPLACE resources that did not yield any pipe items and create empty jobs for them
        # NOTE: do not include tables that have never seen data
        data_tables = {t["name"]: t for t in schema.data_tables(seen_data_only=True)}
        tables_by_resources = utils.group_tables_by_resource(data_tables)
```
looks like this should also go through truncated table state. and other things in `_write_empty_files`... better we fix it after this PR is merged